mlir/test/Quantum/DisentangleCNOTTest.mlir

```diff
@@ -0,0 +1,204 @@
+    // CHECK-NOT: quantum.custom "CNOT"
+    // CHECK: {{%.+}} = quantum.custom "CRZ"({{%.+}}) [[ZERO_0]], [[ONE_1]] : !quantum.bit, !quantum.bit
+
+
+    %5:2 = quantum.custom "CNOT"() %ONE_0, %OTHERS_1 : !quantum.bit, !quantum.bit
+    %user_5:2 = quantum.custom "CRZ"(%cst) %5#0, %5#1 : !quantum.bit, !quantum.bit
+    // CHECK-NOT: quantum.custom "CNOT"
+    // CHECK: [[mid_5:%.+]] = quantum.custom "PauliX"() [[OTHERS_1]] : !quantum.bit
+    // CHECK: {{%.+}} = quantum.custom "CRZ"({{%.+}}) [[ONE_0]], [[mid_5]] : !quantum.bit, !quantum.bit
+
+
+    %6:2 = quantum.custom "CNOT"() %ONE_0, %PLUS_1 : !quantum.bit, !quantum.bit
+    %user_6:2 = quantum.custom "CRZ"(%cst) %6#0, %6#1 : !quantum.bit, !quantum.bit
+    // CHECK-NOT: quantum.custom "CNOT"
+    // CHECK: {{%.+}} = quantum.custom "CRZ"({{%.+}}) [[ONE_0]], [[PLUS_1]] : !quantum.bit, !quantum.bit
+
+
+    %7:2 = quantum.custom "CNOT"() %ONE_0, %MINUS_1 : !quantum.bit, !quantum.bit
+    %user_7:2 = quantum.custom "CRZ"(%cst) %7#0, %7#1 : !quantum.bit, !quantum.bit
+    // CHECK-NOT: quantum.custom "CNOT"
+    // CHECK: {{%.+}} = quantum.custom "CRZ"({{%.+}}) [[ONE_0]], [[MINUS_1]] : !quantum.bit, !quantum.bit
+
+
+    %8:2 = quantum.custom "CNOT"() %ONE_0, %ZERO_1 : !quantum.bit, !quantum.bit
+    %user_8:2 = quantum.custom "CRZ"(%cst) %8#0, %8#1 : !quantum.bit, !quantum.bit
+    // CHECK-NOT: quantum.custom "CNOT"
+    // CHECK: [[mid_8:%.+]] = quantum.custom "PauliX"() [[ZERO_1]] : !quantum.bit
+    // CHECK: {{%.+}} = quantum.custom "CRZ"({{%.+}}) [[ONE_0]], [[mid_8]] : !quantum.bit, !quantum.bit
+
+
+    %9:2 = quantum.custom "CNOT"() %ONE_0, %ONE_1 : !quantum.bit, !quantum.bit
+    %user_9:2 = quantum.custom "CRZ"(%cst) %9#0, %9#1 : !quantum.bit, !quantum.bit
+    // CHECK-NOT: quantum.custom "CNOT"
+    // CHECK: [[mid_9:%.+]] = quantum.custom "PauliX"() [[ONE_1]] : !quantum.bit
+    // CHECK: {{%.+}} = quantum.custom "CRZ"({{%.+}}) [[ONE_0]], [[mid_9]] : !quantum.bit, !quantum.bit
+
+
+    %10:2 = quantum.custom "CNOT"() %PLUS_0, %OTHERS_1 : !quantum.bit, !quantum.bit
+    %user_10:2 = quantum.custom "CRZ"(%cst) %10#0, %10#1 : !quantum.bit, !quantum.bit
+    // CHECK: [[_10:%.+]]:2 = quantum.custom "CNOT"() [[PLUS_0]], [[OTHERS_1]] : !quantum.bit, !quantum.bit
+    // CHECK: {{%.+}} = quantum.custom "CRZ"({{%.+}}) [[_10]]#0, [[_10]]#1 : !quantum.bit, !quantum.bit
+
+
+    %11:2 = quantum.custom "CNOT"() %PLUS_0, %PLUS_1 : !quantum.bit, !quantum.bit
+    %user_11:2 = quantum.custom "CRZ"(%cst) %11#0, %11#1 : !quantum.bit, !quantum.bit
+    // CHECK-NOT: quantum.custom "CNOT"
+    // CHECK: {{%.+}} = quantum.custom "CRZ"({{%.+}}) [[PLUS_0]], [[PLUS_1]] : !quantum.bit, !quantum.bit
+
+
+    %12:2 = quantum.custom "CNOT"() %PLUS_0, %MINUS_1 : !quantum.bit, !quantum.bit
+    %user_12:2 = quantum.custom "CRZ"(%cst) %12#0, %12#1 : !quantum.bit, !quantum.bit
+    // CHECK-NOT: quantum.custom "CNOT"
+    // CHECK: [[mid_12:%.+]] = quantum.custom "PauliZ"() [[PLUS_0]] : !quantum.bit
```

Comment:
This is the first pass that would need to check the supported gate set of the device. For now I think it is sufficient to assume that a user requesting this pass will always run in a device that supports these gates. But one can imagine arriving to this specific transformation and the device not supporting "PauliZ".

Reply:
Is there any action item we can take now regarding this? I think this raises the need to fetch in the supported gateset on a device during mlir passes, and it might be a good thing to have on the technical roadmap.

Reply:
I opened another issue on this, as I think it is a valuable thing to have on the roadmap on its own: https://github.com/PennyLaneAI/catalyst/issues/1394

Reply:
Great remark! We have plans for overhauling the decomposition and extend it to MLIR. Closing this for now.